pinot-tools/src/main/java/com/linkedin/pinot/tools/realtime/provisioning/MemoryEstimator.java

```diff
@@ -0,0 +1,342 @@
+    _consumingMemoryPerHost = new String[numHours.length][numHosts.length];
+
+    int i = 0;
+    for (int numHoursToConsume : numHours) {
+      long minutesToConsume = numHoursToConsume * 60;
+      // consuming for _numHoursSampleSegmentConsumed, gives size sampleCompletedSegmentSizeBytes
+      // hence, consuming for numHoursToConsume would give:
+      long completedSegmentSizeBytes =
+          (long) (((double) minutesToConsume / _minutesSampleSegmentConsumed) * _sampleCompletedSegmentSizeBytes);
+
+      long totalMemoryForCompletedSegmentsPerPartition =
+          calculateMemoryForCompletedSegmentsPerPartition(completedSegmentSizeBytes, numHoursToConsume, retentionHours);
+
+      int totalDocsInSampleSegment = _segmentMetadata.getTotalDocs();
+      // numHoursSampleSegmentConsumed created totalDocsInSampleSegment num rows
+      // numHoursToConsume will create ? rows
+      int totalDocs = (int) (((double) minutesToConsume / _minutesSampleSegmentConsumed) * totalDocsInSampleSegment);
+
+      File statsFileCopy = new File(_tableDataDir, STATS_FILE_COPY_NAME);
+      FileUtils.copyFile(statsFile, statsFileCopy);
+      RealtimeSegmentStatsHistory statsHistory;
+      try {
+        statsHistory = RealtimeSegmentStatsHistory.deserialzeFrom(statsFileCopy);
+      } catch (IOException | ClassNotFoundException e) {
+        throw new RuntimeException(
+            "Exception when deserializing stats history from stats file " + statsFileCopy.getAbsolutePath(), e);
+      }
+      RealtimeIndexOffHeapMemoryManager memoryManager = new DirectMemoryManager(_segmentMetadata.getName());
+      RealtimeSegmentZKMetadata segmentZKMetadata = getRealtimeSegmentZKMetadata(_segmentMetadata, totalDocs);
+
+      RealtimeSegmentConfig.Builder realtimeSegmentConfigBuilder =
+          new RealtimeSegmentConfig.Builder().setSegmentName(_segmentMetadata.getName())
+              .setStreamName(_segmentMetadata.getTableName())
+              .setSchema(_segmentMetadata.getSchema())
+              .setCapacity(totalDocs)
+              .setAvgNumMultiValues(_avgMultiValues)
+              .setNoDictionaryColumns(_noDictionaryColumns)
+              .setInvertedIndexColumns(_invertedIndexColumns)
+              .setRealtimeSegmentZKMetadata(segmentZKMetadata)
+              .setOffHeap(true)
+              .setMemoryManager(memoryManager)
+              .setStatsHistory(statsHistory);
+
+      // create mutable segment impl
+      MutableSegmentImpl mutableSegmentImpl = new MutableSegmentImpl(realtimeSegmentConfigBuilder.build());
+      long memoryForConsumingSegmentPerPartition = memoryManager.getTotalAllocatedBytes();
+      mutableSegmentImpl.destroy();
+      FileUtils.deleteQuietly(statsFileCopy);
+
+      memoryForConsumingSegmentPerPartition += getMemoryForInvertedIndex(memoryForConsumingSegmentPerPartition);
+
+      int j = 0;
+      for (int numHostsToProvision : numHosts) {
```

Comment:
may be better to include the increment of j in the for() statement, so that statements like 'continue' are handled correctly. Something like this:
for (int j = 0; j < numHosts.length; j++) {
  numHostsToProvision = numHosts[j]

Same for the outer for loop

Reply:
Changed